### ISSUE TITLE
Brailliant BI driver does not work over Bluetooth in Windows 10

### ISSUE THREAD

**Issue by nvaccessAuto** [nvaccess/nvda#5195]
**Reported by bramd on 2015-07-03 13:54**
The Brailliant BI driver works fine in Windows 10 as long as the display is connected over USB. However, a connection over Bluetooth is not working. The correct COM port is recognized, but it seems the display is not responding. The traceback indicates the exception at line 127 of brailliantB.py is thrown.

**Comment by jcsteh** [comment]
@DKager, @BramD: I borrowed one of these displays while working on #5612. I wasn't able to reproduce this issue both with and without the changes from #5612. However, #5612 does tighten some things up a bit, so it may have improved things. It's also possible it was improved by a Windows Update; who knows given Microsoft's pathetic lack of documentation on updates. Anyway, unless you can still reproduce this, I think I'll just close this. Note that I don't have the display any more.

**Comment by dkager** [comment]
Connecting the BI 40 through USB worked straight away without custom driver installation, but...
With firmware 2.0.15 on Windows 10 I just got the following traceback while reading with the thumb keys:

```
ERROR - unhandled exception (16:00:36):
Traceback (most recent call last):
  File "_ctypes/callbacks.c", line 314, in 'calling callback function'
  File "hwIo.pyc", line 116, in _ioDone
  File "hwIo.pyc", line 183, in _notifyReceive
  File "serial\serialutil.pyc", line 377, in setTimeout
  File "serial\serialwin32.pyc", line 177, in _reconfigurePort
ValueError: Cannot configure port, some setting was wrong. Original message: [Error 31] A device attached to the system is not functioning.
```

After that my thumb keys stopped working but braille output was unaffected. Choosing the braille display again from the settings fixed it.

**Comment by jcsteh** [comment]
I'm confused. That traceback is serial, not HID, so that means you 
should have needed a custom driver. Or was this exception when connected 
via Bluetooth?

**Comment by dkager** [comment]
This was USB, but with the slightly older firmware that doesn’t yet do HID I think. Maybe Windows 10 knows where to get a driver? It hasn’t happened since yesterday, so seems a rare thing.

**Comment by dkager** [comment]
Just had another instance of the thumb keys ceasing to work, this time without an error being logged. This could be something in the braille display, but I've never seen it with other screen readers. If it keeps happening I'll open a separate ticket.

**Comment by dkager** [comment]
Regarding the serial error in [the comment above](https://github.com/nvaccess/nvda/issues/5195#issuecomment-184246092):
This seems to happen mostly if the system is under load, for example when compiling NVDA, and especially when you rapidly press the thumb keys. Could it be that the serial port is reconfigured by pyserial while the Brailliant still has some output (keypresses) queued? In that case setting a higher timeout value would probably not work.

Switching to the new firmware with HID support should avoid this problem. I'll probably upgrade soon, but wanted to document my findings in case it is of benefit to other serial displays. Would Bluetooth be affected?

**Comment by jcsteh** [comment]
Would you mind filing a separate issue for this exception/input failure? It's definitely not specific to the Brailliant. Thanks.

Btw, unless there's been another firmware release since December, I wouldn't recommend using Brailliant HID just yet. There are serious bugs in that mode (which I've reported to Humanware). Among other things, in HID mode, your display won't charge via USB while powered on. There's also slightly increased latency for key presses in HID mode. And I can't remember the exact details, but I encountered an issue where the display seems to hang when powered down and you have to force reset it. IMO, the QA on that release was utterly atrocious (which I also pointed out to Humanware, if a little more diplomatically).

**Comment by jcsteh** [comment]
Also, are you seeing any issues specific to the Brailliant and Bluetooth now? If not, I think we can close this.

**Comment by dkager** [comment]
> Also, are you seeing any issues specific to the Brailliant and Bluetooth now?

I mainly use a desktop PC without Bluetooth, but I believe @bramd has recent experience.

**Comment by bramd** [comment]
@jcsteh Yes, I used this display over Bluetooth recently. I usually try to prevent this due to latency and sluggishness. However, this seems to be related to the device/firmware and not NVDA specific.

There are a few BT issues with firmware 2.0.15, which are all firmware related. I'll document them here:

1. Bluetooth doesn't seem to work at all in OpenBraille HID mode
2. Bad performance, sluggishness, sometimes display updates quite slowly
3. Pressing one of the space bars when the braille keyboard is turned off in the firmware crashes the display, it will just shut down
4. After a while of inactivity, the display seems to hang and can only be get to work again with a hard reset (holding power button for 15 sec or so), this also happens over USB and also with BRLTTY on Linux, so not NVDA specific. Disabling/enabling sleep doesn't seem to matter for this issue.
5. Display doesn't get recognized all the time on Bluetooth, power cycling and/or going in and out of the firmware config menu sometimes helps. However, this also seems to be a generic problem and not NVDA specific.

In short, yes, I think we can close this issue and hope there will be new firmware that will get some more QA and testing.
